test/compiler/inference.jl

```diff
@@ -3988,3 +3988,21 @@ end |> only == Int
         end
     end |> only === Union{UnionNarrowingByIsdefinedA, UnionNarrowingByIsdefinedB}
 end
+
+# issue #43784
+@testset "issue #43784" begin
+    init = Base.ImmutableDict{Any,Any}()
+    a = Const(init)
+    b = Core.PartialStruct(typeof(init), Any[Const(init), Any, Any])
+    c = Core.Compiler.tmerge(a, b)
+    @test ⊑(a, c)
+    @test ⊑(b, c)
+
+    function f()
+        g = init
+        while true
+            g = Base.ImmutableDict(g, 1=>2)
+        end
+    end
+    @test Base.return_types(f, ()) |> only === Union{}
+end
```

Comment:
```suggestion
@testset "issue #43784" begin
    init = Base.ImmutableDict{Any,Any}()
    a = Const(init)
    b = Core.PartialStruct(typeof(init), Any[Const(init), Any, Any])
    c = Core.Compiler.tmerge(a, b)
    @test ⊑(a, c)
    @test ⊑(b, c)

    @test @eval Module() begin
        const ginit = Base.ImmutableDict{Any,Any}()
        Base.return_types() do
            g = ginit
            while true
                g = Base.ImmutableDict(g, 1=>2)
            end
        end |> only === Union{}
    end
end
```
`init` won't be constant-folded otherwise and thus the last case won't be effective.

Reply:
Good catch!